src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java

```diff
@@ -0,0 +1,289 @@
+		this.partitionKeyExpression = partitionKeyExpression;
+	}
+
+	public void setExplicitHashKey(String explicitHashKey) {
+		setExplicitHashKeyExpression(new LiteralExpression(explicitHashKey));
+	}
+
+	public void setExplicitHashKeyExpressionString(String explicitHashKeyExpression) {
+		setExplicitHashKeyExpression(EXPRESSION_PARSER.parseExpression(explicitHashKeyExpression));
+	}
+
+	public void setExplicitHashKeyExpression(Expression explicitHashKeyExpression) {
+		this.explicitHashKeyExpression = explicitHashKeyExpression;
+	}
+
+	public void setSequenceNumberExpressionString(String sequenceNumberExpression) {
+		setSequenceNumberExpression(EXPRESSION_PARSER.parseExpression(sequenceNumberExpression));
+	}
+
+	public void setSequenceNumberExpression(Expression sequenceNumberExpression) {
+		this.sequenceNumberExpression = sequenceNumberExpression;
+	}
+
+	/**
+	 * Specify a {@link OutboundMessageMapper} for embedding message headers into the record data
+	 * together with payload.
+	 * @param embeddedHeadersMapper the {@link OutboundMessageMapper} to embed headers into the record data.
+	 * @since 2.0
+	 * @see org.springframework.integration.support.json.EmbeddedJsonHeadersMessageMapper
+	 */
+	public void setEmbeddedHeadersMapper(OutboundMessageMapper<byte[]> embeddedHeadersMapper) {
+		this.embeddedHeadersMapper = embeddedHeadersMapper;
+	}
+
+	/**
+	 * Unsupported operation. Use {@link #setEmbeddedHeadersMapper} instead.
+	 * @param headerMapper is not used.
+	 * @see #setEmbeddedHeadersMapper
+	 */
+	@Override
+	public void setHeaderMapper(HeaderMapper<Void> headerMapper) {
+		throw new UnsupportedOperationException("Kinesis doesn't support headers.\n" +
+				"Consider to use 'OutboundMessageMapper<byte[]>' for embedding headers into the record data.");
+	}
+
+	@Override
+	protected Future<?> handleMessageToAws(Message<?> message) throws Exception {
+		if (message.getPayload() instanceof PutRecordsRequest) {
+			throw new UnsupportedOperationException("not implemented");
+		}
+		else {
+			final PutRecordRequest putRecordRequest =
+					(message.getPayload() instanceof PutRecordRequest)
+							? (PutRecordRequest) message.getPayload()
+							: buildPutRecordRequest(message);
+
+			// convert the PutRecordRequest to a UserRecord
+			UserRecord userRecord = new UserRecord();
+			userRecord.setExplicitHashKey(putRecordRequest.getExplicitHashKey());
+			userRecord.setData(putRecordRequest.getData());
+			userRecord.setPartitionKey(putRecordRequest.getPartitionKey());
+			userRecord.setStreamName(putRecordRequest.getStreamName());
+			return handleUserRecord(message, putRecordRequest, userRecord);
+		}
+	}
+
+	private Future<?> handleUserRecord(Message<?> message, final PutRecordRequest putRecordRequest,
+			UserRecord userRecord) {
+		ListenableFuture<UserRecordResult> recordResult = this.kinesisProducer.addUserRecord(userRecord);
+
+		final AsyncHandler<PutRecordRequest, PutRecordResult> asyncHandler =
```

Comment:
I would leave a contract here based on the `UserRecordResult`. A `PutRecordResult`  doesn't bring too much value over the first one and the code here is going to be much simple.

Reply:
As stated in my message, we are not pleased with this UserRecord->PutRecordRequest conversion. It's only there because the AsyncHandler expects an AmazonWebServiceRequest, UserRecord is not.

We can drop this entire mechanism in favor of our own Future handler that mimic the current implementation. We were just not aware of the potential side effects.

Reply:
Wait. I told about `UserRecordResult`, not a `PutRecordRequest`. 
I would prefer to stay with the current `AsyncHandler` approach.

Reply:
`AsyncHandler` and `AbstractAwsMessageHandler` expects an `<R extends AmazonWebServiceRequest>`. We cannot use it with `UserRecord`.

Do you want us to create an `AsyncHandler<PutRecordRequest,UserRecordResult>` ?

We would prefer to use only `UserRecord`/`UserRecordResult` in the entire class (except for payload conversion/retrocompatibility). 
But this asyncHandler is preventing us to do so.